src/App.js

```diff
@@ -606,69 +608,131 @@ export class App extends React.Component<AppProps, AppState> {
-                        window.speechSynthesis.cancel();
+
+                if (matchingKeyboardAction !== false && matchingKeyboardAction !== "partial") {
+                    e.preventDefault();
+                    switch (matchingKeyboardAction) {
+                        case("showHide"):
+                            this.setState((currentState) => {
+                                return { showKeyboardModal: !(currentState.showKeyboardModal) };
+                            });
+                            break;
+                        case("toggleAnnouncements"):
+                            // We have to use the function form here as our change is based on the current state.
+                            this.setState((currentState) => {
+                                return { announcementsEnabled: !(currentState.announcementsEnabled) };
+                            });
+                            break;
+                        case("addCommandToBeginning"):
+                            if (this.state.selectedAction) {
+                                const newProgramSequence = this.state.programSequence.insertStep(0, this.state.selectedAction);
+                                this.handleProgramSequenceChange(newProgramSequence);
+                            }
+                            break;
+                        case("addCommandToEnd"):
+                            if (this.state.selectedAction) {
+                                // $FlowFixMe: Flow doesn't understand that we've already ensured that this.state.selectedAction shouldn't be null.
+                                const newProgramSequence = this.state.programSequence.insertStep(this.state.programSequence.getProgramLength(), this.state.selectedAction);
+                                this.handleProgramSequenceChange(newProgramSequence);
+                            }
+                            break;
+                        case("announceScene"):
+                            const ariaLiveRegion = document.getElementById('character-position');
+                            if (ariaLiveRegion) {
+                                if (window.speechSynthesis.speaking || window.speechSynthesis.pending) {
+                                    window.speechSynthesis.cancel();
+                                }
+                                const utterance = new SpeechSynthesisUtterance(ariaLiveRegion.innerText);
+                                window.speechSynthesis.speak(utterance);
+                            }
+                            break;
+                        case("playPauseProgram"):
+                            if (this.state.programSequence.getProgramLength() > 0) {
+                                this.handleClickPlay();
+                            }
+                            break;
+                        case("refreshScene"):
+                            if (this.state.runningState !== 'running') {
```

Comment:
We want more strict condition check here, as we want to make sure any program execution has stopped. Also, to match when we are disabling the RefreshButton, let's change the condition to match RefreshButton's disabled condition check.

```javascript
if ( !(this.state.runningState === 'stopped' || this.state.runningState === 'paused') ) 
```

Reply:
Done.